src/colvar.cpp

```diff
@@ -1407,6 +1407,10 @@ int colvar::calc()
     error_code |= update_cvc_flags();
     if (error_code != COLVARS_OK) return error_code;
     error_code |= calc_cvcs();
+    if (!cvm::main()->proxy->total_forces_valid()) {
+      // Zero out the colvar total force when atomic total forces are not available
+      ft.reset();
```

Comment:
My impression is that the GPU code doesn't use any of `colvar::calc` and `colvar::calc_cvcs`. The entry point of the GPU code is in `colvarmodule.cpp`:
https://github.com/Colvars/colvars/blob/9f72f0d677f1026d8ad97d302f33c893868296fd/src/colvarmodule.cpp#L1091-L1093
The idea of the GPU code there is flattening the whole tree-like structure in Colvars, constructing a single CUDA graphs from sub-graphs of all atom groups, and then computing all CVCs and their total forces. In other words, the serial code traverses the `colvardeps*` tree by post-order, while the GPU code traverses it by something similar to level-order.